### DISCUSSION TITLE
automaticallyInstallUpdates is deprecated

### DISCUSSION THREAD

**Discussion by colin1994** [sparkle-project/Sparkle#2156] [Q&A] [answered]
> pod 'Sparkle', '2.1.0'

I'm upgrading from Sparkle 1.x to 2.1.0
I found that **installUpdatesIfAvailable** and **automaticallyInstallUpdates** are deprecated.
So, I use **checkForUpdates** instead. (right?

But there is no way to install it automatically, and the "**A new version of %@ is available!**" alert will be displayed.

I want it to show downloading alert directly after checking.
Is this supported? How do I set it up, thanks

**Comment by zorgiepoo** [comment, marked as answer]
No, checkForUpdates is *only* for user-initiated checks.

The supported workflow for automatically downloading/installing updates is using `SUAutomaticallyUpdate ` Info.plist key described here https://sparkle-project.org/documentation/customization/ which installs updates on quit.

[edit]: corrected key

**Reply by colin1994** [reply]
Thank you for your reply, I would like to confirm further.

On my app, there is a button that, when clicked, will attempt to upgrade.
In version 1.x, after the button is clicked, we call **installUpdatesIfAvailable**, which displays a checking alert. Then a downloading alert is displayed. When the download is complete, it will automatically exit and relaunch.

How do I keep this process going on 2.x? (**show alert, auto download, auto relaunch**
Which API to use to trigger the upgrade?
Thanks!

**Reply by zorgiepoo** [reply]
You can have a button that is "Check for updates…" like the typical menu bar item whose action is set to `checkForUpdates` (and bind its `enabled` state to the updater's `canCheckForUpdates`). Then the user can view the release notes and decide to install the update, if there is a new update. Clicking the button is a user-initiated check, so `checkForUpdates` makes sense in this scenario.

In 1.x, using installUpdatesIfAvailable in such a way was buggy; if no new updates were available, it would briefly show and close a checking for updates window.

There isn't a workflow that just automatically visibly shows checking for updates then downloads and then relaunches automatically. Either you have UI interactions which require user input and consent or let Sparkle silently perform checking/downloading/installing in the background to its best ability. (Note sometimes interactions will be required like when an update requires user authorization to install)

**Reply by colin1994** [reply]
Thanks.
Unfortunately, the behavior of 1.x **installUpdatesIfAvailable** is what we most expect.
We would like to have the ability to directly upgrade **remotely**, which does not require user authorization. 
At the same time, I hope that the download progress alert can be displayed on the interface.

But 2.x doesn't seem to support it anymore.

**Reply by zorgiepoo** [reply]
> We would like to have the ability to directly upgrade remotely, which does not require user authorization.

By authorization/authentication I mean if the user needs to elevate permissions to install an update after it's downloaded (like for a standard user account updating an app owned by an admin in /Applications/, for example), which you cannot ensure won't happen.

1.x installUpdatesIfAvailable also allows to cancel the download if I recall correctly.

You can also call `checkForUpdatesInBackground` which will bring up the update only if there's a new one, and do the checking part silently in the background (Sparkle usually schedules this automatically though based on a update check interval but you can also force trigger it).

2.x's standard UI does not support the buggy/unsound installUpdatesIfAvailable workflow 1.x has, sorry. Do note 2.x also provides the ability to write custom user interfaces (via `SPUUserDriver`) if the standard one doesn't meet your needs, which could include automating some of the actions and having a workflow that is more sound.

**Reply by colin1994** [reply]
**1.x**
```
- (void)didFindValidUpdate
{
    if (self.automaticallyInstallUpdates) {
        [self updateAlertFinishedWithChoice:SUInstallUpdateChoice forItem:nil];
        return;
    }
}

- (void)unarchiverDidFinish:(id)__unused ua
{
    if (self.automaticallyInstallUpdates) {
        [self installWithToolAndRelaunch:YES];
        return;
    }
}
```

**automaticallyInstallUpdates** is exactly what we need. （automatically download + installAndRelaunch
Will you consider restoring this parameter later?

If it is no longer supported, is there a way for me to achieve the same effect?

**Reply by zorgiepoo** [reply]
Those were private implementation details, not exposed APIs. Sparkle 2 decoupled the update scheduling and the UI.

I don't know what workflow you're trying to make so I can't give you a good response. For example you can't just have a button (which you vaguely mentioned) that is "install upgrades" in your program because there may not be any updates available. installUpdatesIfAvailable doesn't work properly in 1.x in certain situations and is a hack that really should have never been added. 

You can technically write a custom UI driver (SPUUserDriver) like I already mentioned with a custom user interface to get whatever effect you want. But I can't say this is the right approach because I still don't know the user workflow/context you're trying to make that actually makes sense.

**Reply by zorgiepoo** [reply]
[Quick gist](https://gist.github.com/zorgiepoo/8a1af8cc5ae6878fa10433b3099771b3) to emulate automaticallyInstallUpdates, with warnings that it's not a good idea to do it that way.

**Reply by colin1994** [reply]
Sorry, I didn't describe it properly.
Forget about that **upgrade button**, it doesn't actually exist.

My app, it has no user interaction scene. I don't want it to handle any mouse and keyboard events.
It can only passively accept commands and complete tasks automatically.

So, when the App receives the "**Update**" command from the Server, I hope it can automatically complete this series of tasks: `check --> download --> install -> relaunch `. No user action is required. At the same time, I would like to display the alert(check & download progress bar) to reflect the process.

This is exactly what **installUpdatesIfAvailable** does.

Thanks again for your patient reply!

**Reply by colin1994** [reply]
> [Quick gist](https://gist.github.com/zorgiepoo/8a1af8cc5ae6878fa10433b3099771b3) to emulate automaticallyInstallUpdates, with warnings that it's not a good idea to do it that way.

Technically write a custom UI driver (SPUUserDriver) can simulate this workflow.
Will this remain compatible in the future? Also, **closeCheckingWindow** is a private method, not very friendly.

**Reply by zorgiepoo** [reply]
SPUUserDriver itself is a supported API I make little changes to every once and then but generally try not to break clients. Subclassing SPUStandardUserDriver as you found out is not well supported. So the suggestion would be to write your own UI. This may suit your needs and give you more control anyway, e.g, you can decide not to provide UI cancellation if that's what you want, and also provide smoother animation (as hacking sparkle's standard UI will give you baad results). The UI you want should not be complicated. I assume you are disabling Sparkle's automatic update checks in general and this is your only update workflow.

**Reply by colin1994** [reply]
I get it. Thanks!
